# Final_Project_BIOL7180
This repository includes all the scripts and other files That going to use for the final project of BIOL 7180

## Objective 
The objective of this project is to generate a phylogenetic tree from .fasta sequence files using bash and python scripts

## The Workflow
  - .fasta files of sequences are needed to begin with the project. 
  - These fasta files will be concatenated using a shell command
  - The concateanted fasta files will be aligned using mafft
  - The aligned sequences in .aln format will be manually checked using Mega software
  - The .fas file generated by Mega will be fed to the iqtree command 
  - The resulting .Alignment.fas.treefile will need to be visualized in python
  - The visualization and generation of final tree will be done using python ete3 toolkit
  - A PNG should be visible in the folder after the command has been run


## Preparation

Concatenation of .fasta files using cat function which includes the following command 

<pre>concatenate <font color="#4E9A06">()</font> <font color="#4E9A06">{</font>
  <font color="#3465A4"><b>cat</b></font> *fasta<font color="#4E9A06">&gt;&gt;</font>all_sequence.fasta
<font color="#4E9A06">}</font>
concatenate</pre>

## Alignment 

For alignment, we are going to use mafft which is an open source alignment program that can be used with bash

mafft can be installed with the following command: 
sudo apt install mafft

<pre>sudo apt install mafft </pre>

The following bash function was used to generate an alignment

<pre>align <font color="#4E9A06">()</font> <font color="#4E9A06">{</font> <font color="#06989A">
mafft<font color="#75507B"><b> --localpair --maxiterate</b></font> 1000<font color="#75507B"><b> --clustalout</b></font> all_sequence.fasta<font color="#4E9A06">&gt;</font>Alignment.aln
<font color="#4E9A06">}</font>
align</pre>
